File: backend/retrieve.py
"""
Retrieval module for JLR Technology Intelligence Assistant.
Top-K semantic search with metadata filtering; caps chunks per source to reduce redundancy.
"""
from __future__ import annotations

import json
from pathlib import Path
from typing import Any

from config.load_config import get_project_root, load_config

try:
    import faiss
    import numpy as np
    HAS_FAISS = True
except ImportError:
    HAS_FAISS = False

try:
    from openai import OpenAI
    HAS_OPENAI = True
except ImportError:
    HAS_OPENAI = False


def get_openai_client(api_key: str | None = None, base_url: str | None = None) -> "OpenAI":
    if not HAS_OPENAI:
        raise RuntimeError("openai is required. Install with: pip install openai")
    kwargs: dict = {"api_key": api_key}
    if base_url:
        kwargs["base_url"] = base_url
    return OpenAI(**kwargs)


def load_index_and_metadata(project_root: Path, config: dict[str, Any]) -> tuple[Any, list[dict[str, Any]]]:
    if not HAS_FAISS:
        raise RuntimeError("faiss-cpu is required. Install with: pip install faiss-cpu")
    paths = config.get("paths", {})
    index_dir = project_root / paths.get("index_dir", "index")
    index = faiss.read_index(str(index_dir / "faiss.index"))
    metadata = []
    with open(index_dir / "metadata.jsonl", encoding="utf-8") as f:
        for line in f:
            line = line.strip()
            if line:
                metadata.append(json.loads(line))
    return index, metadata


def embed_query(
    query: str,
    model: str = "text-embedding-3-large",
    client: "OpenAI | None" = None,
    api_key: str | None = None,
) -> list[float]:
    if not client:
        client = get_openai_client(api_key=api_key)
    resp = client.embeddings.create(input=[query], model=model)
    return resp.data[0].embedding


def apply_metadata_filter(
    metadata: list[dict[str, Any]],
    technology_L1: str | None = None,
    technology_L2: str | None = None,
    lifecycle_stage: str | None = None,
) -> set[int]:
    """Return set of indices that pass the filter. Empty filter = all pass."""
    allowed = set(range(len(metadata)))
    if technology_L1:
        allowed &= {i for i, m in enumerate(metadata) if (m.get("technology_L1") or "").lower() == technology_L1.lower()}
    if technology_L2:
        allowed &= {i for i, m in enumerate(metadata) if (m.get("technology_L2") or "").lower() == technology_L2.lower()}
    if lifecycle_stage:
        allowed &= {i for i, m in enumerate(metadata) if (m.get("lifecycle_stage") or "").lower() == lifecycle_stage.lower()}
    return allowed


def retrieve(
    query: str,
    config: dict[str, Any] | None = None,
    *,
    top_k: int | None = None,
    technology_L1: str | None = None,
    technology_L2: str | None = None,
    lifecycle_stage: str | None = None,
    max_chunks_per_source: int | None = None,
    min_similarity_threshold: float | None = None,
    api_key: str | None = None,
) -> list[dict[str, Any]]:
    """
    Run Top-K retrieval with optional metadata filter and per-source cap.
    Returns list of chunk dicts (with 'text' and metadata), ordered by relevance.
    """
    if not HAS_FAISS or not HAS_OPENAI:
        raise RuntimeError("faiss-cpu and openai are required for retrieval.")
    config = config or load_config()
    project_root = get_project_root()
    ret_cfg = config.get("retrieval", {})
    emb_cfg = config.get("embedding", {})

    k = top_k if top_k is not None else ret_cfg.get("top_k", 10)
    max_per_source = max_chunks_per_source if max_chunks_per_source is not None else ret_cfg.get("max_chunks_per_source", 3)
    threshold = min_similarity_threshold if min_similarity_threshold is not None else ret_cfg.get("min_similarity_threshold", 0.0)

    index, metadata = load_index_and_metadata(project_root, config)
    allowed_indices = apply_metadata_filter(metadata, technology_L1, technology_L2, lifecycle_stage)
    if not allowed_indices:
        return []

    base_url = config.get("llm", {}).get("base_url") or None
    client = get_openai_client(api_key=api_key, base_url=base_url)
    model = emb_cfg.get("model", "text-embedding-3-large")
    qvec = embed_query(query, model=model, client=client, api_key=api_key)
    q = np.array([qvec], dtype="float32")
    faiss.normalize_L2(q)

    # Search more than k to allow for per-source capping
    search_k = min(len(metadata), max(k * 3, 50))
    scores, indices = index.search(q, search_k)

    # Filter by allowed indices and threshold; cap per source
    source_count: dict[str, int] = {}
    results: list[dict[str, Any]] = []
    for score, idx in zip(scores[0], indices[0]):
        if idx < 0 or idx not in allowed_indices:
            continue
        if threshold and float(score) < threshold:
            continue
        meta = metadata[idx].copy()
        source = meta.get("source", "")
        if source_count.get(source, 0) >= max_per_source:
            continue
        source_count[source] = source_count.get(source, 0) + 1
        meta["_score"] = float(score)
        results.append(meta)
        if len(results) >= k:
            break

    return results


if __name__ == "__main__":
    import os
    from config.load_config import get_api_key
    config = load_config()
    out = retrieve("What are adoption barriers for digital twins in AEC?", config=config, api_key=get_api_key())
    for r in out:
        print(r.get("chunk_id"), r.get("_score"), r.get("title"))

File: lib/retrieve.js
/**
 * In-memory retrieval: load index (JSON with embeddings + metadata), embed query, cosine similarity, top-k with per-source cap.
 * Index format: { chunks: [ { e: number[], text, source, title, year, authors, doi, apa_citation, chunk_id, ... } ] }
 * or from URL with same shape.
 */
import { getConfig, getApiKey } from './config.js';
import { vercelBlobFetchHeaders } from './vercel-blob-fetch.js';
import OpenAI from 'openai';

const defaultConfig = getConfig();

function dot(a, b) {
  let s = 0;
  for (let i = 0; i < a.length; i++) s += a[i] * b[i];
  return s;
}

function norm(a) {
  return Math.sqrt(dot(a, a)) || 1e-10;
}

function cosineSimilarity(a, b) {
  return dot(a, b) / (norm(a) * norm(b));
}

function l2Normalize(vec) {
  const n = norm(vec);
  return vec.map((x) => x / n);
}

let cachedIndex = null;
let cachedIndexUrl = null;

/**
 * Load index from INDEX_URL (fetch) or INDEX_PATH (file). Cached per process (Vercel serverless reuse).
 */
export async function loadIndex(config = defaultConfig) {
  const url = config.index_url;
  const path = config.index_path;
  if (url) {
    if (cachedIndex && cachedIndexUrl === url) return cachedIndex;
    const res = await fetch(url, { headers: vercelBlobFetchHeaders(url) });
    if (!res.ok) {
      const hint =
        res.status === 401 || res.status === 403
          ? ' For private Vercel Blob, set BLOB_READ_WRITE_TOKEN in env.'
          : '';
      throw new Error(`Failed to fetch index: ${res.status} ${res.statusText}.${hint}`);
    }
    const raw = await res.text();
    const data = JSON.parse(raw);
    cachedIndex = Array.isArray(data.chunks) ? data.chunks : data;
    cachedIndexUrl = url;
    return cachedIndex;
  }
  const fs = await import('fs');
  const pathMod = await import('path');
  const { fileURLToPath } = await import('url');
  const indexFileName = path ? path : pathMod.join('index', 'node_index.json');
  const projectRootFromLib = pathMod.resolve(pathMod.dirname(fileURLToPath(import.meta.url)), '..');
  const candidates = [
    pathMod.resolve(process.cwd(), indexFileName),
    pathMod.resolve(projectRootFromLib, indexFileName),
  ];
  let fullPath = null;
  for (const p of candidates) {
    if (fs.existsSync(p)) {
      fullPath = p;
      break;
    }
  }
  if (!fullPath) {
    throw new Error(
      path
        ? `Index file not found. Tried: ${candidates.join(', ')}`
        : 'No index available. Set INDEX_URL or INDEX_PATH, or place index/node_index.json (run: npm run export-index).'
    );
  }
  if (cachedIndex && cachedIndexUrl === fullPath) return cachedIndex;
  const content = fs.readFileSync(fullPath, 'utf8');
  const data = JSON.parse(content);
  cachedIndex = Array.isArray(data.chunks) ? data.chunks : data;
  cachedIndexUrl = fullPath;
  return cachedIndex;
}

export async function embedQuery(query, model, apiKey, baseUrl) {
  const client = new OpenAI({
    apiKey: apiKey || getApiKey(),
    ...(baseUrl ? { baseURL: baseUrl } : {}),
  });
  const res = await client.embeddings.create({ input: [query], model });
  return res.data[0].embedding;
}

function applyMetadataFilter(chunks, technology_L1, technology_L2, lifecycle_stage) {
  if (!technology_L1 && !technology_L2 && !lifecycle_stage) return chunks.map((_, i) => i);
  const allowed = [];
  for (let i = 0; i < chunks.length; i++) {
    const m = chunks[i];
    if (technology_L1 && (m.technology_L1 || '').toLowerCase() !== technology_L1.toLowerCase()) continue;
    if (technology_L2 && (m.technology_L2 || '').toLowerCase() !== technology_L2.toLowerCase()) continue;
    if (lifecycle_stage && (m.lifecycle_stage || '').toLowerCase() !== lifecycle_stage.toLowerCase()) continue;
    allowed.push(i);
  }
  return allowed;
}

/**
 * Retrieve top-k chunks by semantic similarity. Optional filters; cap per source.
 * Each chunk in index must have .e (embedding) and metadata (text, source, title, ...).
 */
export async function retrieve(query, options = {}) {
  const config = options.config || defaultConfig;
  const apiKey = options.api_key || getApiKey();
  if (!apiKey) throw new Error('OPENAI_API_KEY (or llm.api_key) is required.');

  const top_k = options.top_k ?? config.retrieval.top_k;
  const max_per_source = options.max_chunks_per_source ?? config.retrieval.max_chunks_per_source;
  const threshold = options.min_similarity_threshold ?? config.retrieval.min_similarity_threshold;
  const technology_L1 = options.technology_L1 ?? null;
  const technology_L2 = options.technology_L2 ?? null;
  const lifecycle_stage = options.lifecycle_stage ?? null;

  const chunks = await loadIndex(config);
  if (chunks.length === 0) return [];

  const allowedIndices = applyMetadataFilter(chunks, technology_L1, technology_L2, lifecycle_stage);
  if (allowedIndices.length === 0) return [];

  const embeddingModel = config.embedding?.model || 'text-embedding-3-small';
  const baseUrl = config.llm?.base_url || null;
  const qvec = l2Normalize(await embedQuery(query, embeddingModel, apiKey, baseUrl));

  const withScore = [];
  for (const i of allowedIndices) {
    const c = chunks[i];
    const e = c.e || c.embedding;
    if (!e || e.length !== qvec.length) continue;
    const score = cosineSimilarity(qvec, e);
    if (threshold != null && threshold > 0 && score < threshold) continue;
    withScore.push({ index: i, score, meta: c });
  }
  withScore.sort((a, b) => b.score - a.score);

  const sourceCount = {};
  const results = [];
  // Scan enough candidates to fill top_k across sources (one paper can dominate top scores)
  const searchK = Math.min(withScore.length, Math.max(top_k * 10, 200));
  for (let j = 0; j < searchK && results.length < top_k; j++) {
    const { score, meta } = withScore[j];
    const source = meta.source || '';
    if (sourceCount[source] >= max_per_source) continue;
    sourceCount[source] = (sourceCount[source] || 0) + 1;
    const out = { ...meta };
    delete out.e;
    delete out.embedding;
    out._score = score;
    results.push(out);
  }
  return results;
}
